### PR TITLE
Foods History and Workout History Page

### LINKED ISSUE #47: Page: Workouts report
Show workouts over a given time period, including estimated calories burned, and any other relevant information, for each workout. Statistics such as total calories and average calories per workout should be displayed on this page as well.

### LINKED ISSUE #59: View: food totals
Display all the foods that a user has eaten within a given time, and summary data for each food including total calories, etc for each food.

From the assignment PDF:
> You must implement a page that shows a table that lists all of the food consumed over a selected period of time, along with the calories and macros that are contained in the quantity of food specified. You must implement this page in such a way that clicking on the column headings will make your page sort based on that column. First click should sort in ascending order, a second should sort in descending order, and subsequent clicks just toggle that back and forth.

### LINKED ISSUE #60: Page: Foods report
Display the food totals view within some date range, and allow sorting by any column.

Depends on issue #59 View: food totals

### LINKED ISSUE #77: Bug: setup fails to create views
When I run `setup.sh` it fails on the creating views step, due to this code:
https://github.com/466-mysequel/fitnesstracker/blob/7435b803d43ede94c545865564d5ea35b17d323a/sql/views.sql#L88-L93
Line 89 is causing `setup.sh` to fail because it is missing a matching `DROP VIEW` statement in the `tables.sql` file.

### PR DESCRIPTION
**Pull request by diegoortizb**
- Added a table to view history of your workout activities
- Able to view total and average calories burned
- Added a table to view your food consumption history
- Also added a drop view statement to fix the bug on Issue #77 

 (Side note - I think my code is sloppy, but it works. I am _mentally exhausted_)
 (Side note 2 - I was not able to figure out how to sort with a date range. I don't know if someone wants to pick it up, but I could not figure it out. I spent over 3 hours trying to implement it.)
 (Side note 3 - test out the tables [here](http://students.cs.niu.edu/~z1821368/fitnesstracker/public_html/)

Closes #59 
Closes #47 
Closes #60 
Closes #77

### CODE REVIEW

**Review by J-Erwin — APPROVED** [review]

**Comment by zgjs** [comment]
This looks good. I like the table style and I really like the way you did the sorting by the URL and I think it has a big advantage over the Javascript strategy because as far as I can tell, Javascript only sorts by alphabetical and doesn't sort numbers correctly, while, your MySQL solution is able to correctly sort by numbers. I checked the macros numbers and they are all correct. There is a small problem with some of the columns. The servings number is part of a group, and it is just returning the first value instead of the sum. You can fix it by wrapping `SUM()` around the serving size column. For columns like `date` which are not part of the group and can't be reasonably summed together, it doesn't make sense to include them in the view since it will just return the date of the first food in the group. For that reason, I think any date filtering by date with `WHERE`  needs to happen before the `GROUP BY` clause. I think that means that for the requirement to filter the table by date is not compatible with saving the query as a view. Instead, I think we can either embed the query directly into the page where we want to use it, or use a wrapper function like the standard we have been using for other pages. When we were working together earlier today I wrote the code to wrap the view. It returns an array of associative arrays, so you can use it in any standard result-set processing loop like you would the result of a `PDO::fetchAll()`.
```php
   /**
     * Get food totals
     * 
     * @param user_id The user ID
     * @param start_date A date string in YYYY-MM-DD format to start from
     * @param stop_date A date string in YYYY-MM-DD format to stop from
     */
    function get_food_totals(int $user_id, string $start_date, string $stop_date) {
        $sql = <<<SQL
        SELECT
            f.name AS 'food',
            COUNT(*) AS 'meals',
            SUM(t.servings) AS 'servings',
            f.calories_per_serving AS 'calories_per_serving',
            SUM(t.servings) * f.calories_per_serving AS 'calories',
            CONCAT(SUM(CASE WHEN n.id = 1 THEN mi.amount * t.servings ELSE '0' END), ' g') as 'fat',
            CONCAT(SUM(CASE WHEN n.id = 4 THEN mi.amount * t.servings ELSE '0' END), ' g') as 'carbs',
            CONCAT(SUM(CASE WHEN n.id = 7 THEN mi.amount * t.servings ELSE '0' END), ' g') as 'protein',
            CONCAT(SUM(CASE WHEN n.id = 6 THEN mi.amount * t.servings ELSE '0' END), ' g') as 'fiber'
        FROM food_log t 
        INNER JOIN macronutrient_content mi ON mi.food_id = t.food_id
        INNER JOIN user u ON u.id = t.user_id 
        INNER JOIN food f ON f.id = t.food_id 
        INNER JOIN nutrient n ON n.id = mi.nutrient_id
        WHERE user_id = :user_id AND DATE(t.date) BETWEEN :start_date AND :stop_date
        GROUP BY t.user_id,t.food_id;
        SQL;
        $stmt = $this->pdo->prepare($sql);
        $stmt->execute([
            ':user_id' => $user_id,
            ':start_date' => $start_date,
            ':stop_date' => $stop_date
        ]);
        $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);
        return $rows;
    }
```
For a time range of last 30 days, it should have the numbers:

Food name | Total Meals | Total Servings | Cals / Serving | Total Cals | Total Fat | Total Carbs | Total Protein | Total Fiber
-- | -- | -- | -- | -- | -- | -- | -- | --
bread | 9 | 18 | 75 | 1350 | 2 g | 28 g | 6 g | 4 g
Chik-Fil-A Chocolate Milkshake (small) | 9 | 9 | 610 | 5490 | 22 g | 92 g | 12 g | 1 g
full tomato | 4 | 4 | 30 | 120 | 0 g | 5 g | 1 g | 0 g
large egg | 12 | 30 | 83 | 2490 | 25 g | 5 g | 30 g | 0 g
Lette Caramel Macaron | 7 | 7 | 90 | 630 | 4.5 g | 12 g | 1 g | 0 g
one slice tomato | 7 | 7 | 4 | 28 | 0.042 g | 0.758 g | 0.211 g | 0.195 g
swiss cheese | 14 | 14 | 72 | 1008 | 10 g | 1.688 g | 12 g | 0 g

You can see the output of the function on my branch. The cell contents are correct, but my JavaScript solution doesn't sort numbers properly, so I don't want to use the JavaScript solution.

File: public_html/foods.php [hunk 1]
```diff
@@ -84,21 +84,69 @@
...
-                endforeach;
-            else:
-                echo "        <h3 class=\"mt-3\">Your meal history</h2>\n";
-                $meals = $db->get_meals((int)$_SESSION['user_id']);
-                foreach($meals as $meal):
-                    echo "\n            <h5 class=\"mt-3\"><a href=\"?action=history&timestamp=" . $meal['unixtime'] . "\">". date('h:i:s A l, jS \of F Y', $meal['unixtime']) . "</a></h3>\n";
-                    draw_table($meal['foods'], ['name' => 'Food', 'calories' => 'Calories']);
-                endforeach;
-            endif;
-        break;
```
Comment (zgjs) [inline public_html/foods.php:101]: Please make sure this code does not get removed during a merge.

File: public_html/foods.php [hunk 2]
```diff
@@ -84,21 +84,69 @@
...
+            $sorttype = "asc";
+
+            if(isset($_GET['orderBy']) && $_GET['orderBy'] == $fieldname){
+                if(isset($_GET['sort']) && $_GET['sort'] == "asc"){
+                    $sorttype = "desc";
+                }
+            }
+            $sorturl .= $sorttype;
+            return $sorturl;
+        }
```
Comment (zgjs) [inline public_html/foods.php:107]: This function is really cool. Is it possible to include it in `library.php` so we can reuse it in other places?

File: public_html/foods.php [hunk 3]
```diff
@@ -84,21 +84,69 @@
...
+        if(isset($_GET['orderBy']) && isset($_GET['sort'])){
+            $orderBy = ' ORDER BY '.$_GET['orderBy'].' '.$_GET['sort'];
+        }
+        ?>
+        
+        <?php
+            // HERE WE CHECK USER ID 
+            $user = $db->get_user($_SESSION['user_id']);
+        ?>    
+            <?php $rows = $db->query("select date,food,servings,fat,carbs,protein,fiber from total_food_logs where user_id = ?" . $orderBy, [$_SESSION['user_id']])->fetchAll(PDO::FETCH_ASSOC);?>
```
Comment (zgjs) [inline public_html/foods.php:118]: `$orderBy` contains unsanitized client parameters that could be executed as commands. Those parameters should be added to the PDO prepared statement argument array with only placeholders in the query string.

File: public_html/workouts.php [hunk 4]
```diff
@@ -42,9 +42,85 @@
         </div>
 <?php   break;
         case 'history': ?>
-        <h1>Your workout history</h1>
-        <div class="row">
-        </div>
```
Comment (zgjs) [inline public_html/workouts.php:47]: This is OK to delete if you need to.

File: sql/tables.sql [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # Cleanup:
 \! echo "Dropping tables and views if they exist..."
-DROP VIEW IF EXISTS micro_totals_today, micro_totals_weekly, micro_totals_monthly, micronutrients, macronutrients, macro_totals_monthly, macro_totals_weekly, macro_totals_today, net_calories_per_day, calories_in_per_day, calories_out_per_day, current_weight, workout_calories_burned;
+DROP VIEW IF EXISTS micro_totals_today, micro_totals_weekly, micro_totals_monthly, micronutrients, macronutrients, macro_totals_monthly, macro_totals_weekly, macro_totals_today, net_calories_per_day, calories_in_per_day, calories_out_per_day, current_weight, workout_calories_burned, total_food_logs;
```
Comment (zgjs) [inline sql/tables.sql:3]: 👍

File: sql/views.sql [hunk 6]
```diff
@@ -87,9 +87,20 @@ GROUP BY user_id,nutrient_id;
...
-INNER JOIN macronutrient_content USING (food_id)
-GROUP BY user_id,food_id ORDER BY date ASC;
+select t.user_id,
+       f.name as 'Food',
+       t.servings as 'servings',
+       SUM(CASE WHEN n.id = 1 THEN mi.amount * t.servings ELSE '0' END) as 'fat',
+       SUM(CASE WHEN n.id = 4 THEN mi.amount * t.servings ELSE '0' END) as 'carbs',
+       SUM(CASE WHEN n.id = 7 THEN mi.amount * t.servings ELSE '0' END) as 'protein',
+       SUM(CASE WHEN n.id = 6 THEN mi.amount * t.servings ELSE '0' END) as 'fiber',
+       t.date as 'Date'
```
Comment (zgjs) [inline sql/views.sql:97]: This query is good and works with all macros. Servings and calories should also be `SUM()`;ed. `date` is not part of the group-by clause and cannot be summed, so it should be removed.

**Review by zgjs — CHANGES_REQUESTED** [review]
Hi @diegoortizb thanks for submitting this PR! The table sorting works really well and I want to include it on the main landing page for foods, and I want to re-use your sorting function on other pages if that is ok. Are you able to modify it to work with prepared statements and save the function to one of the library files under `/src/`?

**Review by HR0102 — APPROVED** [review]
Looks Good but before you merge just make the changes that Greg asked you to make

**Comment by diegoortizb** [comment]
@zgjs, feel free to make any changes you want to this branch. I can no longer work on it. I have other priorities.

File: sql/tables.sql (hunk 5 above)
Comment (z1868762) [inline sql/tables.sql:3]: closes #77

**Review by z1868762 — APPROVED** [review]
image

**Pull request merged**